account_reconcile_oca/models/account_bank_statement_line.py

```diff
@@ -0,0 +1,750 @@
+        to_reconcile = defaultdict(lambda: self.env["account.move.line"])
+        with move._check_balanced(container):
+            for line in suspense_lines | other_lines:
+                to_reconcile[line.account_id.id] |= line
+                line_data = line.with_context(
+                    active_test=False,
+                    include_business_fields=True,
+                ).copy_data({"move_id": move.id})[0]
+                to_reconcile[line.account_id.id] |= (
+                    self.env["account.move.line"]
+                    .with_context(
+                        check_move_validity=False,
+                        skip_sync_invoice=True,
+                        skip_invoice_sync=True,
+                    )
+                    .create(line_data)
+                )
+            move.write(
+                {
+                    "line_ids": [
+                        Command.update(
+                            line.id,
+                            {
+                                "balance": -line.balance,
+                                "amount_currency": -line.amount_currency,
+                            },
+                        )
+                        for line in move.line_ids
+                        if line.move_id.move_type == "entry"
+                        or line.display_type == "cogs"
+                    ]
+                }
+            )
+            for line_vals in data:
+                if line_vals["kind"] == "liquidity":
+                    continue
+                if line_vals["kind"] == "suspense":
+                    raise UserError(_("No supense lines are allowed when reconciling"))
+                line = (
+                    self.env["account.move.line"]
+                    .with_context(check_move_validity=False, skip_invoice_sync=True)
+                    .create(self._reconcile_move_line_vals(line_vals, move.id))
+                )
+                if line_vals.get("counterpart_line_id") and line.account_id.reconcile:
+                    to_reconcile[line.account_id.id] |= (
+                        self.env["account.move.line"].browse(
+                            line_vals.get("counterpart_line_id")
+                        )
+                        | line
+                    )
+            move.invalidate_recordset()
+        move._post()
+        for _account, lines in to_reconcile.items():
+            lines.reconcile()
+
+    def unreconcile_bank_line(self):
+        self.ensure_one()
+        return getattr(
+            self, "_unreconcile_bank_line_%s" % (self.reconcile_mode or "edit")
+        )()
+
+    def _unreconcile_bank_line_edit(self):
+        self.reconcile_data_info = self._default_reconcile_data(from_unreconcile=True)
+        self.action_undo_reconciliation()
+
+    def _unreconcile_bank_line_keep(self):
+        raise UserError(_("Keep suspense move lines mode cannot be unreconciled"))
+
+    def _reconcile_move_line_vals(self, line, move_id=False):
```

Comment:
because in the function, it used `self.move_id.id` which will not work when `self` contains multiple records.

we need to use `ensure_one()` when the function involves operations with recordsets and it should only operate on a single record

So it might be:
```python
if not move_id:
    self.ensure_one()
```

Reply:
No need to ensure_one on private functions. That should be managed on public functions.